optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/drools/common/AbstractAccumulator.java

```diff
@@ -85,6 +60,37 @@ public final ResultContainer_ init(Object workingMemoryContext, Object context,
         return containerSupplier.get();
     }
 
+    @Override
+    public final Object accumulate(Object workingMemoryContext, Object context, Tuple leftTuple,
+            InternalFactHandle handle, Declaration[] declarations, Declaration[] innerDeclarations,
+            ReteEvaluator reteEvaluator) {
+        /*
+         * Accumulator instances are created within the KieBase, not within the KieSession.
+         * This means that sessions may be calling this method on a shared accumulator from different threads.
+         * Unfortunately, we need access to innerDeclarations to be able to perform the accumulation quickly,
+         * and it only becomes available when this accumulate(...) method is called.
+         * Therefore, initialization of our accumulator can only happen the first time the method is called,
+         * and therefore the initialization needs to be properly synchronized.
+         *
+         * We do not care if the initialization happens multiple times,
+         * as each of the attempts will fill the variables with the same data.
+         * We just need to ensure that the variables are initialized before the actual accumulation happens.
+         *
+         * Technically running this more than once is wasteful,
+         * but the initialized() check will make sure this only happens during the cold start of the session.
+         */
+        if (!initialized.get()) {
+            initialize(leftTuple, innerDeclarations);
+            initialized.set(true);
+        }
+        return accumulate((ResultContainer_) context, leftTuple, handle, innerDeclarations);
+    }
+
+    protected abstract Runnable accumulate(ResultContainer_ context, Tuple leftTuple, InternalFactHandle handle,
+            Declaration[] innerDeclarations);
+
+    protected abstract void initialize(Tuple leftTuple, Declaration[] innerDeclarations);
```

Comment:
This is similar to different to Accumulator.init(). Can we add a comment here to explain why init() doesn't furfill our needs? Something like "// WORKAROUND because init() doesn't do X"

Reply:
The comment is already there, and it is very long.